torch/csrc/api/include/torch/data/datasets/base.h

```diff
@@ -0,0 +1,67 @@
+#pragma once
+
+#include <torch/data/example.h>
+#include <torch/detail/utils.h>
+#include <torch/tensor.h>
+
+#include <ATen/core/ArrayRef.h>
+
+#include <cstddef>
+#include <cstdint>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace torch {
+namespace data {
+namespace datasets {
+template <typename S, typename T>
+struct Map;
+template <typename D, typename T>
+Map<D, T> map(D&&, T&&);
+} // namespace datasets
+} // namespace data
+} // namespace torch
+
+namespace torch {
+namespace data {
+namespace datasets {
+
+template <typename S, typename B = std::vector<Example<>>>
+class BatchDataset {
+ public:
+  using Self = S;
+  using BatchType = B;
+
+  virtual ~BatchDataset() = default;
+
+  virtual BatchType batch(ArrayRef<size_t> indices) = 0;
+  virtual size_t size() const = 0;
+
+  template <typename TransformType>
+  Map<Self, TransformType> map(TransformType transform) && {
+    return datasets::map(
+        std::move(static_cast<Self&>(*this)), std::move(transform));
+  }
+};
+
+template <typename S, typename E = Example<>, typename B = std::vector<E>>
+class Dataset : public BatchDataset<S, B> {
+ public:
+  using typename BatchDataset<S, B>::BatchType;
+  using ExampleType = E;
+
+  virtual ExampleType index(size_t index) = 0;
```

Comment:
Both this and `batch` could be overloaded instances of `operator[]`, so datasets would look like regular containers.

Reply:
The reason I don't use operator overloads, whether its `operator[]` for datasets or `operator()` for transforms, is that we need two overloads:
1. One for batches,
2. One for examples (which will often call the batch version).
Having two overloads makes inheritance less friendly, because of name hiding rules in C++. Basically, if you want to override one in your own subclass, you'd be hiding the other. So every subclass would need a line like `using BatchDataset<...>::operator()`.

Instead, I just give the two overloads different names. That solves the problem. It's a little less sweet, but much easier on users.